Implementation of model type

KangKyung opened the pull request:
# 배경

앱 개발에 대한 전반적인 느낌은, [기능명세📝](https://github.com/KangKyung/ios-market/blob/main/FunctionalRequirements.md)를 참고해주세요!
이번 PR은 앱의 Model 타입, **네트워크 Model과 JSON매핑 모델 구현**을 중점적으로 다뤘습니다!

# 작업 내용

- [ ] #2 
- [ ] #3 
- [ ] #4 

# 테스트 방법

이번 PR은 GUI환경에서 테스트하기가 어려워요😢
그 대신 `MarketTests/NetworkModel`에 만든 단위테스트를 활용해주세요!
단위테스트를 진행하면서 동작에 불필요한 코드는 없는지, 어색한 로직은 없는지 확인부탁드릴게요🙇🏻‍♂️

# 리뷰 노트

### 네트워크에 의존하지 않는 단위테스트
 
저는 `MockURLProtocol`을 이용하여, 네트워크의존성을 역전시켜 외부에서 주입을 받도록 만들었습니다.
혹시나, 더 나은 방법을 알고있다면 알려주세요‼️

### 리뷰 가이드

이번 PR에서는 목적별 기능추가를 주로 다뤘기 때문에
전체파일을 한번에 리뷰하는것 보다는, 커밋단위로 따라가시는게 편할거에요!

# 스크린 샷

- 혹시나, 이 외에 더 진행해보면 좋을거 같은 테스트가 있다면 알려주세요😎
<img width="361" alt="스크린샷 2021-09-02 오후 4 34 49" src="https://user-images.githubusercontent.com/38858863/131802171-e4f87db0-b4cb-44f5-95fc-a3973870c32b.png">

KangKyung committed:
Init Project

KangKyung committed:
Chore: Packaging

KangKyung committed:
Chore: SwiftLint 설정

KangKyung committed:
[#2]Feat: 파싱한 JSON데이터와 매핑할 모델 생성

KangKyung committed:
[#3]Feat: Network 연결모듈 생성

KangKyung committed:
[#4]Test: Network Model 단위테스트 진행

YongHoonJJo commented on Market/Market/Sources/Constants/HttpMethod.swift:
```diff
@@ -0,0 +1,38 @@
+    
+    case get, post, put, patch, delete
+    
+    var description: String {
```
`CustomStringConvertible` 은 해당 객체의 description을 정의할 때 필요한 protocol 로 알고 있습니다. 
`URLRequestProtocol` 의 `makeURLRequest()` 메서드에서 `httpMethod.description` 와 같이 사용을 하고 계신데요.

[https://swiftdoc.org/v4.2/protocol/customstringconvertible/](https://swiftdoc.org/v4.2/protocol/customstringconvertible/) 

위 문서에서는 `Accessing a type's description property directly or using CustomStringConvertible as a generic constraint is discouraged.` 라고 설명하고 있습니다. 이부분에 대해서는 어떻게 생각하시는지 궁금합니다.!

KangKyung replied:
> CustomStringConvertible 은 해당 객체의 description을 정의할 때 필요한 protocol 로 알고 있습니다.

저도 의견에 동감합니다!! 기존 레거시코드를 리팩토링하는 과정에서 미처 찾지못한 부분이었어요😅
생각 못했던 부분을 짚고 넘어가주셔서 감사합니다🙏🏼

Wody95 replied:
첫번째 코멘트와 마찬가지로 `description`에서 `switch`문으로 반환하는 `String`이 특별한 패턴을 보이지 않고 있으므로 `CustomStringConvertible`은 채택하지 않아도 될 것 같아보입니다. 

강경은 어떻게 생각하시나요? ☺️

KangKyung replied:
저도 Jost와 우디의 의견에 동의합니다! 저땐 왜 저렇게 코드를 짰을까요..🛌🦶🏼

YongHoonJJo commented on Market/Market/Sources/Model/Response/ProductSearchResponseModel.swift:
```diff
@@ -0,0 +1,42 @@
+    let stock: Int
+    let discountedPrice: Int?
+    let thumbnails: [String]
+    let images: [Data]
```
API 명세에서는 String Array 로 본 것 같은데요. 혹시 제가 잘못알고 있거나, 아니면 다른 의도가 있던 걸까요?

KangKyung replied:
해당 모델은 POST요청시 `Request Body`로 들어갈 때 사용합니다!
API문서의 `Response Body`에서는 말씀대로 String Array타입이라고 명시되어있지만
`Request Body`에서는 File Array타입이라고 명시되어 있어서, 일단 Data 타입으로 만들었어요🔨
제가 아직 그 이후 기능개발까지 진행해보지 않아서그런데, request 진행시 string으로 보내야하는걸까요..?

YongHoonJJo replied:
Request 시에는 API 명세를 따르면 될거에요. multipart-form 형식으로 해서 다른 데이터들과 함께 바이너리 채로 요청을 받는 경우도 존재할 수도 있고, 이미지만 따로 서버에 먼저 바이너리 형태로 보내고, 이미지가 저장된 스토리지의 주소를 받은 경우, [String] 으로 보내면 되기 때문에 아마 multipart-form 형식으로 보내진 않아도 될거에요. 하지만 저희 명세에는 전자의 방식을 따르니 말씀하신 것 처럼 [Data] 형태로 진행하면 될 것 같아요. 

다만 제가 의문을 제기했던 이유는 파일명이 ResponseModel 이라고 되어 있어서 URLSession 의 Response 에 오는 data 를 파싱하기 위한 모델이라고 생각이 들어서 문의드렸었습니다. 또한 요청을 보낼때는 섬네일을 같이 보내진 않을 텐데, 모델에는 섬네일에 대한 프로퍼티까지 포함을 하고 있어서 문의드렸었습니다!!

KangKyung replied:
@YongHoonJJo 
... 죄송합니다 Jost의 말이 맞았어요😢
이거 `POST`때 쓰는 모델이 아니라 `GET`할 때 쓰는 모델이 맞습니다..
진짜 디테일한 피드백 감사합니다 ㅠㅠ 저보다 Jost가 제코드를 더 잘파악 하셨네요... 부끄럽습니다😭

YongHoonJJo replied:
@KangKyung 
죄송할거까지 있나요 ㅎㅎㅎㅎ 홧팅입니다!!
STEP2 도 진행하시면 DM 이나 채널에 노티 부탁드리겠습니다 :)

YongHoonJJo commented on Market/Market/Sources/Constants/RequestType.swift:
```diff
@@ -0,0 +1,51 @@
+        case .loadProduct(let id):
+            return "/item/\(id)"
+        case .postProduct:
+            return "/item/"
```
여기만 슬래시로 끝나는 것 같아요!

YongHoonJJo commented on Market/Market/Sources/Constants/RequestType.swift:
```diff
@@ -0,0 +1,51 @@
+    var url: URL? {
+        return URL(string: "\(RequestType.baseURL)\(urlPath)")
+    }
+    var httpMethod: HttpMethod {
```
이 프로퍼티는 활용되는 곳이 보이질 않는 것 같아요. 
테스트 코드에서 `makeURLRequest(httpMethod: .get, apiRequestType: apiRequestType)` 이렇게 사용되는 코드가 있는데요, `.get` 대신 `apiRequestType.httpMethod` 와 같이 사용 가능할 것 같은데 어떻게 생각하시나요?

KangKyung replied:
이 역시 기존 레거시코드를 리팩토링하는 과정에서 미처 찾지못한 부분이군요..😥
그런데 코드의 변경에대한 제 생각은 조금 다릅니다..!
`var httpMethod`자체가 `enum RequestType`에서 불필요한거같아요
해당 역할은 이미 `enum HttpMethod`에서 충분히 해주고있는 것 같거든요!
가령, '앞으로 `loadPage`를 진행할 때는 `get`이아니라 `post`를 사용해야 겠군!'과 같은 상황에서는
전체 코드를 고칠필요없이 해당 부분만 수정하면 되니까 유용할 수도 있겠지만
사실, 그럴일은 없을거 같거든요😅 혹시, Jost는 이 부분에 대해 어떻게 생각하시나요??

YongHoonJJo replied:
`var httpMethod` 자체가 `enum RequestType` 에서 불필요한 것이라면 괜찮습니다!! 다만 해당 코드가 존재하기에, 해당 코드를 제가 제안드린 방식으로 사용하기 위한 의도로 만든것이 아니였을까란 생각에 문의드려봤습니다 :)

YongHoonJJo commented on Market/Market/Sources/Network/Protocol/MultiPartRequestProtocol.swift:
```diff
@@ -0,0 +1,59 @@
+        
+        return urlRequest
+    }
+    func multiPartBody(name: String, value: Any) -> Data {
```
아래의 `multiPartBodyOfImage()` 메서드와 같이 위의 `setMultiPartBody()` 메서드 내부에서만 사용되는 메서드 같은데요. 접근 제어를 추가하면 어떨까 생각합니다. 혹시 다른 이유가 있었다면 의견 부탁드리겠습니다!

KangKyung replied:
저도 습관처럼 `private`이나 `final`, `let`을 사용하는 걸 지향합니다만,
변명하기도 창피한 부분이군요😭 그러게 왜 안썼을까요.......
꼼꼼한 리뷰 감사드립니다🙇🏻‍♂️

YongHoonJJo commented:
안녕하세요. 코드 잘 봤습니다! 저는 이 플젝을 진행하면서 네트워킹 부분을 이렇게 구조적으로 분리를 하지 않았던 탓이 이렇게 분리를 할 수도 있구나 하고 생각하고, 또 배울 수 있었습니다. 제가 드린 코멘트는 단지 저의 생각일 뿐이니 편하게 의견에 대해 답해주시면 감사하겠습니다!!


추가로 의견을 드리자면,

#### RequestType.swift
`RequestType` 타입 내의 `urlPath` 과 `httpMathod` 프로퍼티의 경우, 같은 값을 리턴하는 부분에 대한 case 를 합치는 것에 대해서는 어떻게 생각하는 궁금합니다. 
RequestType 타입의 경우 item 이란 리소스에 대한 CRUD 에 맞게 설계가 되어 보이는데, 다른 리소스들에 대한 CRUD 가 추가되면 case 계속 추가되어야 하는 구조로 보여서 이렇경우 나중에 case 에 따라 라인이 많이 질 것 같단 생각이 들었습니다. 혹은, 처음부터 그런 부분을 염두해둔 설계하신 것인지가 궁금합니다!


#### URLRequestProtocol.swift / MultiPartRequestProtocol.swift
아직은 많이 부족한 상태라 그런지, 개인적으로 protocol + extension 으로의 분리가 조금은 과해보인단 생각이 듭니다. ㅜ
위의 두 소스파일의 구현 내용에 대해 protocol + extension 으로 둔 이유가 있으실까요?
구현체에 대한 공통 메서드를 MarketAPIProvider 타입 이외에서 필요로 하다면 위와 같은 것이 좋은 선택이라고 생각합니다만, 현재 구조에서는 MarketAPIProvider 이외에서는 필요로 하는 곳이 있을까란 생각이 들었습니다. ㅜ 이부분에 대해 제가 미처 생각하지 못한 부분이 있다면 의견 부탁드리겠습니다 :)

stevenkim18 commented on Market/Market/Sources/Model/Response/Item.swift:
```diff
@@ -0,0 +1,38 @@
+//
+
+import Foundation
+
+struct Item: Decodable {
+    
```
다른 모델과 다르게 이 친구만 이름이 Item 인 이유는 뭔가요?

KangKyung replied:
```swift
struct ListSearchResponseModel: Decodable {
    
    let page: Int
    let items: [Item]
}
```
상품을 리스트로 받아오는 JSON 매핑 모델 타입이 위와같아서
무의식 적으로 items라는 이름에 맞게 Item이라고 지었어요!
그러게 스티븐 말 듣고 생각해보니, Product라는 이름이 더 어울릴거 같기도 하네요~?

stevenkim18 commented on Market/Market/Sources/Model/Response/Item.swift:
```diff
@@ -0,0 +1,38 @@
+        case registrationDate = "registration_date"
+    }
+    
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.id = try container.decode(Int.self, forKey: .id)
+        self.title = try container.decode(String.self, forKey: .title)
+        self.price = try container.decode(Int.self, forKey: .price)
+        self.currency = try container.decode(String.self, forKey: .currency)
+        self.stock = try container.decode(Int.self, forKey: .stock)
+        self.discountedPrice = try container.decode(Int?.self, forKey: .discountedPrice)
+        self.thumbnails = try container.decode([String].self, forKey: .thumbnails)
+        self.registrationDate = try container.decode(Double.self, forKey: .registrationDate)
+    }
```
이니셜라이즈를 만드신게 `discountedPrice` 가 옵셔널이라서 그런가요?

KangKyung replied:
네... 정확합니다😢
Decodable을 채택하니까 이니셜라이저를 반드시 등록해야해서요...
혹시 스티븐은 다른 해결방법을 알고계신가요??

Wody95 replied:
저는 강경과 똑같이 `struct`에 Decodable을 채택했는데 이니셜라이즈를 등록하지 않아도 디코딩이 가능했습니다. 무슨 차이가 있는걸까요?🧐

KangKyung replied:
우디도 `discountedPrice`를 옵셔널 타입으로 만들었는데 디코딩이 가능했던건가요..? 저만 뭐가문제인지 모르나보군요😭

stevenkim18 replied:
저도 정확한건 아닌데 타입이 한두개 없어도 디코딩이 가능한 경우를 봤어요.. ㅎㅎ 저도 한번 테스트를 해봐야할것 같은데.. ㅎㅎ

stevenkim18 replied:
아아 그리고 리뷰 못남긴 것이 있는데 모델들이 중복되는 것들이 너무 많아서 공통되는 부분들은 프로토콜로 뺴도 좋을 것 같아요 ㅎㅎ

KangKyung replied:
> 저도 정확한건 아닌데 타입이 한두개 없어도 디코딩이 가능한 경우를 봤어요.. ㅎㅎ 저도 한번 테스트를 해봐야할것 같은데.. ㅎㅎ

아니 그러게... 지금 또 해보니까 되네요.. 이게 왜 되는거지......
분명 안되서 이니셜라이즈 찾아가지고 한건데😭

> 모델들이 중복되는 것들이 너무 많아서 공통되는 부분들은 프로토콜로 뺴도 좋을 것 같아요 ㅎㅎ

좋은 포인트네요! 감사합니다☺️

stevenkim18 commented on Market/Market/Sources/Network/MarketAPIProvider.swift:
```diff
@@ -0,0 +1,97 @@
+//
+
+import Foundation
+
+struct MarketAPIProvider {
+    
```
클래스 말고 구조체로 구현하신 이유가 있을까요?

KangKyung replied:
제가 상속을 지양하는 편이고, 굳이 class로 구현할 필요가 없으면 struct를 사용하는 편이에요!
혹시 스티븐은 이 부분을 class로 구현했었나요??
사실 해당부분에 대해 깊은고민을 해봤던건 아니라서, 스티븐은 어떤 기준으로 class와 struct를 사용할지 고민하시나요?😲

stevenkim18 replied:
APIProvider가 여기저기서 굉장히 많이 쓰일 것 같아서.. 그때마다 인스턴스를 만드는 것은 조금 비효율적이라고 생각해서 저는 클래스로 선언해서 싱글톤으로 만들어요 ㅎㅎ 그러면 인스턴스 한개로 여러번 사용할 수 있잕아요.. 

그런데 생각해보니 싱글톤은 구조체에서도 사용할 수 있는 것 같네요...

저도 왠만하면 구조체를 쓰려고 하고 참조가 꼭 필요한 경우에만 클래스를 쓰려고 해요 ㅎㅎ

stevenkim18 commented on Market/Market/Sources/Network/Protocol/MultiPartRequestProtocol.swift:
```diff
@@ -0,0 +1,59 @@
+        guard var urlRequest = makeURLRequest(httpMethod: httpMethod,
+                                              apiRequestType: apiRequestType) else { return nil }
+        var data = Data()
+        let mirror = Mirror(reflecting: product)
+        mirror.children.forEach { key, value in
```
`Mirror`를 사용하셨네요 ㅎㅎ 이유가 궁금합니다. 컴파일 타임이 아닌 런타임에서 작동 된다고 들었는데 성능에는 큰 문제가 없을까요?

KangKyung replied:
request를 진행할 때, image를 어떻게 전달해야 할지 고민하다가 사용하게되었어요!
API에서 요구하는 image의 타입이 File Array (저는 이 부분을 [Data]라고 해석했습니다.)이라고 기재되어있어서
데이터 형식으로 변환해주어야 하는데, 방법을 찾지 못하겠더군요..
그래서 성능 부분은 전혀 생각지도 못했어요😢
혹시, 스티븐은 해당 부분에대해 어떤 고민을 하셨고, 어떤 생각을 가지고 있는지 매우매우 궁금합니다..!!

stevenkim18 replied:
저도 다시 구현을 해봐야 알곘지만.. 저도 오픈 마켓 할 때 미러 사용을 고려 했다가 저는 그냥 모델에서 연산프로퍼티로 그냥 하드코딩을 했었어요 ㅎㅎ

그 이유는
전에 글을 찾아보다가 컴파일 타임과 런타임 타임의 차이점에 대해서 잠깐 보게 되었는데.. 컴파일 타임에서 애러가 나는 경우는 그냥 xcode에서 빌드 자체가 안되는데 런타임에서 오류는 앱이 실행중에 날 수 있다고 봤어요! 물론 그럴 가능성은 아주 적겠지만 `Mirror` 는 런타임에서 실행되는 코드라고 들어서 예상치 못한 오류가 날 수 있을거라고 생각해서 여쭤본거에요 ㅎㅎ

저도 다시 컴파일과 런타임의 차이를 찾아봤는데 
https://dd-corp.tistory.com/9
이런 글이 있네요 ㅎㅎ 뭐 검색해보면 많이 나올 것 같아요 ㅎㅎ 

그리고 미러를 거의 사용하지 않는 것 같아서 그럴 만한 이유가 있을거라고 생각했는데.. 이건 더 고수분에게 물어봐야될것 같네요 ㅎㅎ

stevenkim18 reviewed (commented):
강경 코드 잘 봤습니다.. 전반적으로 코드가 저희가 예전에 했던 거랑은 큰 틀에서 벗어나지 않은 것 같아요 ㅎㅎ
그래서 그냥 코드를 보면서 왜 이렇게 구현하셨는지 궁금한 부분만 코멘트 남겼습니다.

사실 테스트에 대해서도 저도 아직은 잘 모르겠어요.. ㅎㅎ 강경처럼 `URLProtocol` 을 사용해서 더미 객체를 넘겨주는 테스트가 의미 있는 테스트인지 궁금해요 ㅎㅎ 혹시 모르면 같이 알아가요 ㅎㅎㅎ

KangKyung commented:
@YongHoonJJo 
정말 너무나도 꼼꼼히 봐주셔서 놀랐어요...
리뷰 하나하나마다 많은 시간을 할애한, 많은 고민이 느껴졌습니다👍🏻
그런만큼 저의 실수가 적나라하게 드러나서 창피했지만😅 덕분에 더욱 단단한 프로젝트가 된거 같아요!

## 추가 의견에 대한 답변

> RequestType 타입 내의 urlPath 과 httpMathod 프로퍼티의 경우, 같은 값을 리턴하는 부분에 대한 case 를 합치는 것에 대해서는 어떻게 생각하는 궁금합니다.

item이 아닌 다른 리소스들에 대한 CRUD가 추가될 때를 염두하신걸로 이해했습니다
확실히 설계를 하면서 생각해보지 못했던 부분이었어요..!
음.. 하지만 그럴경우에는 몇가지 문제가 있을거 같다는 생각이 들었어요
```swift
private var urlPath: String {
    switch self {
    case .loadProduct(let id):
        return "/item/\(id)"
    case .patchProduct(let id):
        return "/item/\(id)"
    }
}
```
예를 들어 위의 경우를 보면, 같은 값을 리턴하는 두 경우는 `read`와 `update`인데
1. 가독성 측면에서 어떤 네이밍으로 합쳐야할 지 모르겠고
2. 나중에 patchProduct의 url path를 변경하게 된다면, 합쳐져있는 loadProduct를 다시 분리하게 될거 같아요

그래서 제 생각에는 현재의 구조를 유지하되, 다른 리소스를 추가하게되는 경우
**다른 리소스들에 대한 `RequestType`을 더 만들어서 해결**하는게 좋다는 생각이드는데, 어떨까요??

> protocol + extension 으로 둔 이유가 있으실까요?

현재의 구조는 `Request Body`의 처리에 대해 고민하다가 나온 결과입니다!
이 부분의 코드를 처음 구성할 때는 `Protocol`에 대한 이해가 얕았을 때였고, 리팩토링을 진행할 때 까지도 별 생각이 없었지만
한참동안 Jost의 질문을 곰곰히 생각해보니 개선할 필요성을 느꼈어요‼️
개선뒤에 다시 말씀드릴게요! 날카롭고 멋진 포인트였어요ㅎㅎ 감사합니다🙇🏻‍♂️

Wody95 reviewed (commented):
안녕하세요 강경 😊

강경의 코드를 보면서 URLSession API를 어떻게 구성해야 할지 고민한 흔적이 보여 리뷰하기 즐거운 코드였습니다! 만 다른 분들이 워낙 섬세하게 질문해주셔서 제가 남길 코멘트가 몇 없네요 ㅠㅠ

저는 강경에게 드리고 싶은 질문은 네트워크와 무관한 상태에서, 서버에서 받은 데이터를 표현하고 사용할 수 있는 테스트 코드에 대해 질문드리고 싶습니다.

제가 이해한 강경의 테스트 코드는 네트워크 상황을 상정하거나, 그에 따라 테스트를 진행하는 것으로 보여 질문 드립니다 😊

KangKyung commented:
@stevenkim18 
역시 스티븐.. 제가 구현하기 힘들었던 부분만 골라서 질문하셔서 답변이 늦었습니다...
답변이 바로바로 안나오고 늦어진다는건, 결국 제가 해당 부분에대해 많이 모르기 때문인거 같아요😢
그렇기에, 스티븐은 해당 부분에 대해 어떤 생각이 들었고, 어떤 고민들을 했을지가 궁금했어요!
그러니까 위 코멘트들에 꼭 답변해주세요~~ ㅎㅎㅎ

> URLProtocol 을 사용해서 더미 객체를 넘겨주는 테스트가 의미 있는 테스트인지 궁금해요

데이터를 MockURLProtocol을 이용해서 (실제로는 넘어가지 않지만)서버로 넘어가는 과정이 알맞는지 확인하는게 포인트라고 생각해요!
해당 부분을 테스트 하는데에 더미를 이용하는부분이 적합한건지 궁금하셨던 걸까요?
조금만 더 구체적으로 질문해주시면, 더 명확한 답변이 가능할거 같아요!

그런데, 생각을 정리해보려고 다시 제 테스트코드를 다시 확인했는데
이부분을 좀 비몽사몽한 상태로 해서 엉망이더군요😅
현재는 테스트 코드 전에 네트워크 요청에 대한 성공여부를 지정해주고 이에따른 성공여부만 체크하고있는데
정말 의미없는 테스트라서... parameter등이 알맞게 요청되는지 확인하는 느낌으로 변경하려고 해요!
혹시, 이 부분 때문에 그렇게 말씀하신거라면.. 제가 이상한 테스트를 진행한게 맞습니다😫

KangKyung commented:
@Wody95 

오! 우디!! 정말 궁금했던 테스트 부분에 대해 코멘트 남겨주셔서 감사합니다😁
우디 말대로 어떻게 구조를 설계해야할지 많은생각을했고, 특히나 Testable한 코드를 만들기위해 고민했던거 같아요!
하지만, 노력에도 여전히 미숙한 코드였던거 같아 부끄럽습니다😅

> 저는 강경에게 드리고 싶은 질문은 네트워크와 무관한 상태에서, 서버에서 받은 데이터를 표현하고 사용할 수 있는 테스트 코드에 대해 질문드리고 싶습니다.

> 제가 이해한 강경의 테스트 코드는 네트워크 상황을 상정하거나, 그에 따라 테스트를 진행하는 것으로 보여 질문 드립니다 😊

리뷰받기 전까지는 나름 완벽하다고😎 느끼는데, 왜 항상 부끄러운 코드는 리뷰를 받고나서 보이는 걸까요😭
우디 말대로, 현재 제 테스트는 네트워크 요청에 대한 성공여부를 지정해주고 이에따른 성공여부만 체크하고있어요...
해당 부분을 좀 더 개선해보도록 해보겠습니다! 감사합니다🙇🏻‍♂️

KangKyung committed:
[#5]Refactor: 피드백 반영

- struct Item 네이밍 변경 -> Product
- enum HttpMethod CustomStringConvertible채택 취소
- protocol MultiPartRequestProtocol 일부함수 private
- enum RequestType
  - urlPath 오타 수정
  - httpMethod 메서드변수 삭제

KangKyung commented:
- [ ] #6

YongHoonJJo commented:
```swift
private var urlPath: String {
    switch self {
    case .loadProduct(let id):
        return "/item/\(id)"
    case .patchProduct(let id):
        return "/item/\(id)"
    }
}
```

요 부분은

```swift
private var urlPath: String {
    switch self {
    case .loadProduct(let id), .patchProduct(let id):
        return "/item/\(id)"
    }
}
```

이런식으로 case 를 합쳐보는 것이 어떤지에 대해 의견을 물어봤었습니다!!
API 명세가 바뀌면 그부분은 어쩔수 없이 수정을 해야한다고 생각은 합니다.  :)

전반적으로 리뷰에 대해 긍정적으로 반응해주셔서 감사합니다!!

KangKyung committed:
[#6]Test: 네트워크에 의존하지않는 테스트 개선

KangKyung commented:
@YongHoonJJo 아하... 케이스를 합친다는게 이걸 말했던거군요!!😅 이건 완전 찬성이죠👍🏻

KangKyung committed:
[#5]Refactor: 피드백 반영

- RequestType.urlPath 중복되는case 하나로 통일

KangKyung committed:
Test: 불필요한 try-catch구문 삭제
